File: 2015/day01a.go
package main

import (
	"fmt"
	"os"
	"log"
	"io/ioutil"
)

func printusage() {
	fmt.Printf("Usage: %s <directions file>\n\nFind Santa's location based on up and down directions given in <directions file>\n", os.Args[0])
	log.Fatal("FATAL: no filename given")
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func main() {
	if len(os.Args) != 2 {
		printusage()
	}

	dat, err := ioutil.ReadFile(os.Args[1])
	check(err)

	floor := 0
	for i := 0; i < len(dat); i++ {
		switch(dat[i]) {
			case '(':
				floor++
			case ')':
				floor--
		}
	}
	fmt.Printf("Santa should arrive on floor %v\n", floor)
}

File: 2015/day01a.c
#include <stdio.h>
#include <stdlib.h>

char * readfile(const char *fname) {
	FILE * f;
	size_t s, t;
	char * data;
	f = fopen(fname, "r");
	if (NULL == f) {
		fprintf(stderr, "There was an error opening the input file.  Aborting!\n");
		exit(1);
	}
	if (fseek(f, 0L, SEEK_END)) {
		fprintf(stderr, "Error seeking to end of file.  Aborting!\n");
		exit(1);
	}
	s = ftell(f);
	rewind(f);
	data = malloc(s+1);
	if (NULL == data) {
		fprintf(stderr, "Could not allocate memory for file.  Aborting!\n");
		exit(1);
	}
	t = fread(data, sizeof(char), s, f);
	if (t != s) {
		fprintf(stderr, "Could not read file.  Aborting!\n");
		exit(1);
	}
	data[t+1] = '\0';
	return(data);
}

int main(int argc, char **argv) {
	char * filedata, * dir;
	int floor = 0;
	if (argc != 2) {
		printf ("Please provide the filename containing your puzzle input as the first argument\n");
		return 1;
	}
	filedata = readfile(argv[1]);
	dir = filedata;
	while (*dir) {
		switch (*dir++) {
			case '(':
				floor++;
				break;
			case ')':
				floor--;
		}
	}
	free(filedata);
	printf ("Santa should arrive on floor %d\n", floor);
}
